UnicodeMode support for regexp

balajirrao opened the pull request:
Adds partial support for 'u' regexp flag. 

- **Only Unicode property-escapes that are on the JDK are supported**: full support would need icu4j. One possibility is to have a `rhino-icu4j` jar that can offer full unicode support. I'll create a discussion.
- **the 'i' flag is not supported in UnicodeMode**: doing this needs case-folding handling for property-escapes. See the second paragraph [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Unicode_character_class_escape#description).
- **Unicode property-escape tests are disabled**: Although a large percentage of the Unicode property-escape test262 tests pass with this PR, they're still disabled due to the tests taking forever to run. All of them use regexps like `/\p{Foo}+$/` to match on long strings. These regexps setup backtracking for each character in the input. I was able to make these regexps go faster with a quick-and-drity hack, by avoiding backtrack setup. Going to work on it soon again.

rbri commented:
Violations also present in:
      src/test/java/org/mozilla/javascript/tests/NativeRegExpTest.java
  Run './gradlew :rhino:spotlessApply' to fix these violations.

rbri commented:
Great
![image](https://github.com/user-attachments/assets/e7b425e7-094d-4090-bda1-223c53974787)

zbynek commented on rhino/src/main/java/org/mozilla/javascript/regexp/UnicodeProperties.java:
```diff
@@ -0,0 +1,437 @@
+                    Map.entry("Script", SCRIPT),
+                    Map.entry("sc", SCRIPT),
+                    Map.entry("Hex_Digit", HEX_DIGIT),
+                    Map.entry("Hex", HEX_DIGIT),
```
The way it's implemented corresponds to "AHex", Ascii hex digit.
`\p{Hex}` should also match e.g. https://www.compart.com/en/unicode/U+FF11

zbynek commented on rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java:
```diff
@@ -1112,19 +1195,206 @@ private static boolean parseIdentityEscape(CompilerState state, ParserParameters
+    }
+
+    // only in the format \p{X} or \P{X}. Assume the \ has been consumed.
+    // depending on p or P choose PROP_ESC or PROP_NO_ESC
```
did you mean "`REOP_UPROP` or `REOP_UPROP_NOT`"?

zbynek commented on rhino/src/main/java/org/mozilla/javascript/regexp/UnicodeProperties.java:
```diff
@@ -0,0 +1,437 @@
+                return Character.isUpperCase(codePoint) == (valueByte == TRUE);
+
+            case WHITE_SPACE:
+                return Character.isWhitespace(codePoint) == (valueByte == TRUE);
```
Unicode's notion of whitespace maps to `Character.isSpaceChar`, not `isWhiteSpace`, because of different handling of control characters and non-breaking spaces.

`"\u00A0".match(/\p{space}/u)`

zbynek commented:
I was just working on this mapping in the opposite direction (making GWT [translate Unicode checks](https://github.com/gwtproject/gwt/pull/10113/files) from Java to JS). It would be great to have this merged to Rhino to be able to test the Java->JS->Java round-trip. Thanks for working on this! I left a few minor comments for consideration.

balajirrao commented:
> I was just working on this mapping in the opposite direction (making GWT [translate Unicode checks](https://github.com/gwtproject/gwt/pull/10113/files) from Java to JS). It would be great to have this merged to Rhino to be able to test the Java->JS->Java round-trip. Thanks for working on this! I left a few minor comments for consideration.

@zbynek thanks for your comments, they're very useful - I've not really worked with Unicode before.

I've been busy for the last couple of weeks. I'll aim to get back to this next week and have it ready soon!

zbynek commented:
@balajirrao sorry for nagging, any progress on this? I managed to run some GWT tests against a build of Rhino including these changes and the result was very promising.

balajirrao commented:
Hey @zbynek, just a quick heads-up — still working through this! It's taking me a bit longer than I expected since I had to refresh some of the context from the PR. Thanks for your patience — I’ll get back with updates soon.

balajirrao commented:
@zbynek I've addressed your comments and marked the PR ready for review. Returning to the PR after three months proved to be more challenging than I had anticipated. 😅

balajirrao commented:
Hi @gbrail, just a quick note to let you know that this PR is ready for review. It’s been a draft for more than 3 months now, so I wanted to make sure it’s on your radar. Thanks!

rbri commented:
@balajirrao is is ok to squash on merge?

rbri commented on rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java:
```diff
@@ -409,6 +412,17 @@ private static void prettyPrintRE(RECompiled regexp) {
+                    pc += INDEX_LEN;
+                    char lowSurrogate = (char) getIndex(regexp.program, pc);
+                    pc += INDEX_LEN;
+                    System.out.println(
```
a left over? or something we have to hide behind a debug flag?

rbri commented:
LGTM - anybody else?

rbri commented:
> @balajirrao is is ok to squash on merge?

After checking this again i think i will merge that without squashing if nothing else comes up until the end of today.

balajirrao commented:
@rbri yes, let's squash on merge.

rbri commented:
Thanks a lot!
